salt/states/pyenv.py

```diff
@@ -216,4 +216,10 @@ def install_pyenv(name, user=None):
         ret['comment'] = 'pyenv is set to be installed'
         return ret
 
-    return _check_and_install_python(ret, user)
+    ret =  _check_pyenv(ret, user)
+    if not ret['result'] is True:
```

Comment:
This would be better written:

```suggestion
    if not ret['result']:
```

No need to check against `True`. Also, `is` is an identity comparison.

Reply:
Thanks for the review!

Yeap, philosophically you are right, but in practice the result must be equal only  the True, not any non empty string, not a number, etc. And such situation is possible in any non-typed language.

Unfortunately, I'm not familiar with the project test system. Could you please add some dummy test to the PR, so I'll add a real test.

Reply:
Philosophically, you're also right, but practically speaking Salt internals should (MUST?) return `True`, `False`, or `None` for the result.

In fact this code is trivial enough to follow - https://github.com/saltstack/salt/blob/8eecd9c69f41c2ef227318f73c228d0f541f58cb/salt/states/pyenv.py#L57

You're setting `result` to `None`, and it can only ever be set to `False` in that function, which actually makes me wonder what this test is even meant to check? It will never ever be `True` or anything even remotely truthy, until after this line, if I'm reading this correctly.

Reply:
> Philosophically, you're also right, but practically speaking Salt internals should (MUST?) return `True`, `False`, or `None` for the result.

 The whole file is _not_ my code.  I just added as short as possible patch. I do not understand who must or should?
> 
> In fact this code is trivial enough to follow -
> 
> https://github.com/saltstack/salt/blob/8eecd9c69f41c2ef227318f73c228d0f541f58cb/salt/states/pyenv.py#L57
> 
> You're setting `result` to `None`, and it can only ever be set to `False` in that function, which actually makes me wonder what this test is even meant to check? It will never ever be `True` or anything even remotely truthy, until after this line, if I'm reading this correctly.

 It's NOT my code.

Anyway, thank you, this's a good note. I forgot to add another part of patch ))) Another part returns True in case pyenv is installed.

Generally speaking, nobody knows what other function will return.

Reply:
https://docs.saltstack.com/en/latest/ref/states/writing.html#return-data

> result: A tristate value. True if the action was successful, False if it was not, or None if the state was run in test mode, test=True, and changes would have been made if the state was not run in test mode.

We shouldn't worry if `result` is ever anything but `True`, `False`, or `None`. 

```
$ ag 'if ret\[.result.\]:' -c
tests/unit/states/test_module.py:1
salt/cloud/clouds/lxc.py:1
salt/states/memcached.py:2
salt/states/elasticsearch_index_template.py:1
salt/states/elasticsearch.py:5
salt/states/elasticsearch_index.py:1
salt/states/user.py:1
salt/states/webutil.py:1
salt/states/selinux.py:1
salt/states/win_path.py:2
salt/states/cron.py:1
salt/states/ports.py:1
salt/states/logrotate.py:2
salt/states/virt.py:1
salt/states/at.py:1
salt/states/zfs.py:9
salt/states/win_system.py:1
salt/states/win_pki.py:2
salt/states/reg.py:1
salt/states/http.py:1
salt/states/zone.py:12
salt/states/zpool.py:4
salt/states/group.py:1
salt/states/file.py:3
salt/states/logadm.py:2
salt/states/rbenv.py:2
salt/states/mdadm_raid.py:2
salt/states/grains.py:2
salt/states/pyenv.py:1
salt/states/smartos.py:5
salt/states/jboss7.py:1
salt/modules/nagios_rpc.py:2
salt/modules/dockermod.py:1
```

There *are* other places in the code where we use `is True`:

```
$ ag 'if ret\[.result.\] is True:' -c
salt/states/lxc.py:1
salt/states/heat.py:2
salt/states/timezone.py:1
salt/modules/heat.py:2
```

Out of those, I'd say that only one of those files *might* be following a reasonable pattern, but even that one still feels a bit odd.

> Generally speaking, nobody knows what other function will return.

My understanding is that any salt functions that are returning a `'result'` value outside of `True, False, or None` should be considered buggy and fixed.

Reply:
Just for interest. Could you please also grep for 'if.*is True'?

Reply:
It's a first time in my life when somebody states that implicit check is better than explicit.

Reply:
https://docs.quantifiedcode.com/python-anti-patterns/readability/comparison_to_true.html

Very *very* rarely will you see Python code checking for *absolute* "Truth". As a matter of fact, I've seen more `is True` in Salt that I've seen since Python 2.4!

In other languages you check for equality with true and false, but Python code has a strong preference for relying on the boolean nature of things, because it's rare that you need to be so explicit. And I would argue that it means that you wrote your code wrong.

Reply:
1) I have used not a comparison but the strong check!
2) 
> I've seen more is True in Salt that I've seen since Python 2.4!
because Salt was the devops tool (and now I'm doubt if it will stay as it is and not become a garbage of the states). We can suggest everything that we want, but if it's possible to go by a wrong way it will definitely go by the wrong way. So an explicit declaration is always preferable (despite most people are lazy and don't use it).
3) in other languages they also have special mechanisms for the strong comparison with True or False. And because of many languages (especially nontyped) have such mechanism I think they know something )))